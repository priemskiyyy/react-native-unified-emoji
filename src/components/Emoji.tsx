import React from 'react';
import { Image, type ImageProps, type ImageSource } from 'expo-image';
import { getEmojiCdnPath } from '../utils/getEmojiCdnPath';
import { EmojiSet } from '../constants/set';
import { useImageStatus } from '../utils/useImageStatus';
import NativeEmoji from './NativeEmoji';

export type EmojiProps = {
  children: string;
  source?: ImageSource;
  set?: EmojiSet;
  Fallback?: typeof NativeEmoji;
  Loader?: React.FunctionComponent<Pick<EmojiProps, 'children'>>;
} & Omit<ImageProps, 'children' | 'source'>;

const Emoji: React.FunctionComponent<EmojiProps> = ({
  children,
  source: _source = {},
  Fallback = NativeEmoji,
  Loader = () => null,
  set = EmojiSet.Apple,
  ...imageProps
}) => {
  const { status, ...handlers } = useImageStatus(imageProps);

  if (!children) {
    return null;
  }

  const uri = getEmojiCdnPath(children, set);

  const source: ImageSource = {
    ...(_source ?? {}),
    uri: uri,
    cacheKey: uri,
  };

  if (status === 'error') {
    return <Fallback children={children} />;
  }

  return (
    <>
      <Image {...imageProps} {...handlers} source={source} />
      {status === 'loading' && <Loader children={children} />}
    </>
  );
};

export default Emoji;
