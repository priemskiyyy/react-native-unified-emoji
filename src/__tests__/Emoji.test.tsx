import Emoji from '../components/Emoji';
import { render } from '@testing-library/react-native';

jest.mock('expo-image', () => ({
  Image: 'Image',
}));
describe('Emoji', () => {
  it('renders correctly', () => {
    const { toJSON, getByTestId } = render(<Emoji testID={'emoji'}>ðŸ‘‹</Emoji>);
    expect(toJSON()).toMatchSnapshot();
    expect(getByTestId('emoji')).toBeTruthy();
  });
  it('not crash when children is empty', () => {
    // @ts-ignore
    const { toJSON } = render(<Emoji />);
    expect(toJSON()).toMatchSnapshot();
  });
});
