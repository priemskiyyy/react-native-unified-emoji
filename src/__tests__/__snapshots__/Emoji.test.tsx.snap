// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emoji not crash when children is empty 1`] = `null`;

exports[`Emoji renders correctly 1`] = `
<Image
  onError={[Function]}
  onLoad={[Function]}
  onLoadStart={[Function]}
  source={
    {
      "cacheKey": "https://emojicdn.elk.sh/ðŸ‘‹?style=apple",
      "uri": "https://emojicdn.elk.sh/ðŸ‘‹?style=apple",
    }
  }
  testID="emoji"
/>
`;
